src/engine-actions/translate-swap.js

```diff
@@ -38,40 +50,141 @@ async function translateSwap (takerAddress, swapHash, amount, maxTime, finalCltv
+      throw new PermanentSwapError(`Timelock for total swap is shorter than final hop of payment. ` +
+        `cltvLimt: ${cltvLimit}, finalCltvDelta: ${finalCltvDelta}`)
+    }
+
+    const {
+      paymentPreimage,
+      paymentError
+    } = await sendPayment({
+      paymentHash: swapHash,
+      destString: publicKey,
+      amt: amount,
+      finalCltvDelta,
+      cltvLimit,
+      feeLimit: DEFAULT_FEE_LIMIT
+    }, { client: this.client })
+
+    if (paymentError) {
+      this.logger.error(`Payment error while sending to route`, { paymentError, swapHash })
+      throw new PermanentSwapError(paymentError)
+    }
+
+    return paymentPreimage
   }
+}
+
+/**
+ * Retrieve the preimage for the given payment hash. If no payment is found
+ * for the hash, it will throw a gRPC NOT FOUND error. If the payment is
+ * in-flight, it will wait until the payment succeeds or is in a permanent
+ * failure state.
+ * @param {string} paymentHash - Base64 encoded payment hash
+ * @param {LndClient} client
+ * @param {Logger} logger
+ * @returns {Promise<string>} - Base64 encoded string of the preimage
+ */
+async function getPreimageIfPaymentExists (paymentHash, client, logger) {
+  return new Promise((resolve, reject) => {
+    try {
+      if (!paymentHash) {
+        throw new Error('paymentHash must be defined')
+      }
+
+      logger.debug('Attempting to track payment for paymentHash:', { paymentHash })
+      const stream = trackPayment(paymentHash, { client })
+
+      const errorListener = (err) => {
+        logger.error('Error from trackPayment stream', { paymentHash })
+        stream.removeAllListeners()
+        return reject(err)
+      }
+
+      const endListener = () => {
+        logger.error('LND closed trackPayment stream before returning our value')
+        stream.removeAllListeners()
+        return reject(new Error('LND closed stream to track payment'))
+      }
+
+      const dataListener = (response) => {
+        const { state, preimage } = response
+
+        if (state === PAYMENT_STATUSES.IN_FLIGHT) {
+          // trackPayment will continue to stream updates of the payment state, so
+          // no action is required on our end
+          return logger.debug('Payment is In-Flight for paymentHash:', { paymentHash })
+        }
+
+        if (state === PAYMENT_STATUSES.SUCCEEDED) {
+          if (!preimage) {
+            stream.removeAllListeners()
+            return reject(new Error('No preimage associated with successful payment'))
+          }
+
+          logger.debug('Payment has Succeeded for paymentHash', { paymentHash })
+          stream.removeAllListeners()
+          return resolve(preimage)
+        }
+
+        if (state === PAYMENT_STATUSES.FAILED_TIMEOUT || state === PAYMENT_STATUSES.FAILED_NO_ROUTE) {
+          logger.error('Payment failed', { paymentHash })
+
+          stream.removeAllListeners()
+          return reject(new PermanentSwapError(`Payment failed for paymentHash: ${paymentHash}`))
```

Comment:
might be nice to include some info about the failure (`TIMEOUT` or `NO_ROUTE` at a minimum, more info if LND provides it)

Reply:
LND will only provide `PaymentState` on failures, so I'll add the `state` to logging.